Packages: Add notices package

aduth opened the pull request:
Closes #6388

This pull request seeks to extract notices behaviors from the `editor` module to a new `notices` module, intended to serve largely the same purpose, but generalized to serve use-cases outside the editor. The API also largely remains the same, except with the addition of a new `context` option for specifying a simplified "area" value for notices grouping.

Planned to fork out into separate pull requests:

- _Already: #9615. Removed from changes here since no longer in active use after iterations._
- Failing unit tests are because we don't load dependent stores within tests. When running tests within a directory, we should probably include its top-level `index.js` to ensure these dependencies have their expected stores defined. See related #8911.
- I copied `onSubKey` higher-order reducer creator from `core-data`. I'm starting to wonder if we might want to consider distributing these as some form of data/reducer utilities, akin to [as `@wordpress/compose` is for higher-order components](https://github.com/WordPress/gutenberg/tree/master/packages/compose). Maybe a `@wordpress/compose-data` or `@wordpress/compose-reducer`.

Open question:

- With introduction of "context", should we consider most of the notices as we have today as part of an "editor" context, or does the idea of a global context represent what we're showing today in the editor below the heading? It'd be my idea that what we have today serves the "global" usage, but it is screen-specific, and we might want to consider "properly global" notices (things like [notifications center](https://core.trac.wordpress.org/ticket/43484)), where our notices aren't quite as global.
- What should be included here so far as componentry? The "Notice" component seems generic enough that it's sensible to remain in `@wordpress/components`. With this module effectively serving as an orchestrator, it seems most all of the state-bound components should live within.

Future enhancements:

- Our existing notices should ideally not include React elements, and rather represent content as a plain string. Most of our existing use of elements could be better served by a formal "actions" option (e.g. "View Post" on the save success) which is more future-compatible to notices design revisions.
- I'm curious to explore whether the new `context` concept could be leveraged to eliminate most of what's served by the internal state of [`withNotices`](https://github.com/WordPress/gutenberg/blob/master/packages/components/src/higher-order/with-notices/index.js).

gziolo commented on packages/notices/package.json:
```diff
@@ -0,0 +1,34 @@
+{
+	"name": "@wordpress/notices",
+	"version": "1.0.0",
```
`1.0.0-beta.0` please :)

gziolo commented on packages/notices/package.json:
```diff
@@ -0,0 +1,34 @@
+	],
+	"main": "build/index.js",
+	"dependencies": {
+		"@babel/runtime-corejs2": "7.0.0-beta.56",
```
Is should be replaced with:
```json
"@babel/runtime": "^7.0.0",
```
since we are now using Babel 7 stable.

aduth commented:
@gziolo @youknowriad I'm curious your opinion on:

>I copied `onSubKey` higher-order reducer creator from `core-data`. I'm starting to wonder if we might want to consider distributing these as some form of data/reducer utilities, akin to [as `@wordpress/compose` is for higher-order components](https://github.com/WordPress/gutenberg/tree/master/packages/compose). Maybe a `@wordpress/compose-data` or `@wordpress/compose-reducer`.

Another option I'm contemplating is repurposing `@wordpress/compose` to serve for any higher-order function composition, made evident by its `compose` import which would otherwise need to be duplicated redundantly into this new package if one were to be created. I'm not sure if it might cause some confusion however to have all these functions as part of the same package, partly for bloat and partly because it becomes unclear whether an import is to serve the purpose of creating a higher-order component or a higher-order reducer. Granted, there already is some confusion already in that some of what we export is technically a "higher-higher-order-function" (higher-order component creators).

aduth committed:
Packages: Add notices package

Co-Authored-By: Grzegorz (Greg) Ziółkowski <grzegorz@gziolo.pl>

aduth committed:
Notices: Add actions support

Makes spokenMessage redundant by ensuring content is to be only assigned as plain string

Co-Authored-By: Grzegorz (Greg) Ziółkowski <grzegorz@gziolo.pl>

aduth commented:
This is now "complete" and could do for a proper review.

gziolo commented:
> >I copied `onSubKey` higher-order reducer creator from `core-data`. I'm starting to wonder if we might want to consider distributing these as some form of data/reducer utilities, akin to [as `@wordpress/compose` is for higher-order components](https://github.com/WordPress/gutenberg/tree/master/packages/compose). Maybe a `@wordpress/compose-data` or `@wordpress/compose-reducer`.
> 
> Another option I'm contemplating is repurposing `@wordpress/compose` to serve for any higher-order function composition, made evident by its `compose` import which would otherwise need to be duplicated redundantly into this new package if one were to be created. I'm not sure if it might cause some confusion however to have all these functions as part of the same package, partly for bloat and partly because it becomes unclear whether an import is to serve the purpose of creating a higher-order component or a higher-order reducer. Granted, there already is some confusion already in that some of what we export is technically a "higher-higher-order-function" (higher-order component creators).

I don't find it as a big problem that we duplicate 10 lines of code. We could land it as is. However if I would have to vote where to put this helper function, I think it makes more sense to make `@wordpress/compose` a general purpose component which offers all tools which help with composition. The only drawback I see is that this package depends on `@wordpress/element`. 

> I put together a rough concept for what I'd considered to be improvements to the test helpers to accommodate for the need to ensure stores for package are loaded (by ensuring all package src/index.js are imported for test). It had some negative cascading effects, so I opted instead for a more simplified approach here (basically import '../../' to trigger index imports).

It seems like a nice improvement to the current experience if you assume that those stores should be initialized. Personally, I think the ugliness of those imports raises the question whether our code doesn't depend too much on the internal state of stores. Maybe we should be more cautious when introducing such dependencies. Maybe we should rethink the way we mock and verify the usage of the data store. For instance, we could create mock helper utility which would allow us to easily operate on stores:
```js
const noticesStore = mockStore( 'core/notices', {
    actions: {
        createSuccessNotice: jest.fn(),
    }
} );

expect( dataDispatch( 'core/notices' ).createSuccessNotice ).not.toHaveBeenCalled();
```
This way you wouldn't have to import `@wordpress/notices` package at all.

Another question is how often we could offer a low-level pure function which operates on data fetched from the store which could be much easier unit tested.

gziolo commented:
I added https://github.com/WordPress/gutenberg/pull/9617/commits/e9e89384e7aa493d845a337b75197d1e51bc7073 since `token-list` is used only in tests.

gziolo commented on packages/editor/src/store/actions.js:
```diff
@@ -536,49 +537,6 @@ export function stopTyping() {
-			status,
-			content,
-			isDismissible,
-			spokenMessage,
```
This code assumes that from now on `spokenMessage` is always equal to `content` - is it correct?

aduth replied:
I believe so, in that previously we'd allow `spokenMessage` to be of type `element`, which is no longer supported (must be string).

gziolo replied:
Yeah right. It makes stnyyo consolidate now.

gziolo commented on packages/notices/src/store/selectors.js:
```diff
@@ -0,0 +1,55 @@
+ *
+ * @type {Array}
+ */
+const DEFAULT_NOTICES = [];
```
Nit: Should we also move to `constants.js` file?

aduth replied:
Maybe. I think the main value in `constants.js` are for variable constants used across files. This one is specific to selectors, however.

aduth replied:
> Nit: Should we also move to `constants.js` file?

I think I'd rather keep it here, per my previous comment.

gziolo reviewed (commented):
I think the only part left to review is all the changes in store effects. Otherwise it all looks good.

gziolo commented on packages/notices/src/store/utils/on-sub-key.js:
```diff
@@ -0,0 +1,30 @@
+ *
+ * @return {Function} Higher-order reducer.
+ */
+export const onSubKey = ( actionProperty ) => ( reducer ) => ( state = {}, action ) => {
```
My second thought is that it might be also a good idea to move it to `data` package next to other helpers like `combineReducers` as it has a very similar application. Eventually, we could extract those utilities to their own package. It's very unlikely that we will use it without reducers :)

aduth replied:
I think I like that, especially if even the main fear would be bloat for a size-conscious module developer, it'd be easily omitted by tree shaking.

aduth replied:
> I think I like that, especially if even the main fear would be bloat for a size-conscious module developer, it'd be easily omitted by tree shaking.

Let's keep this to a follow-up pull request.

gziolo replied:
Agreed, this one is big enough 👍

gziolo commented on packages/notices/src/store/test/actions.js:
```diff
@@ -0,0 +1,188 @@
+			} );
+		} );
+
+		it( 'should return action with context', () => {
```
Super nit: we can make this statement more explicit - `...action with *custom* context`

aduth replied:
> Super nit: we can make this statement more explicit - `...action with *custom* context`

Updated in rebased 009b8e916b3ae9fac32af65261713feab454097f.

gziolo commented on packages/notices/src/store/reducer.js:
```diff
@@ -0,0 +1,36 @@
+		case 'CREATE_NOTICE':
+			// Avoid duplicates on ID.
+			return [
+				...reject( state, { id: action.notice.id } ),
```
Oh nice, I was using `filter( not( ... ) )` - it's a nice shortcut and reads much better 👍

gziolo commented on packages/notices/src/store/actions.js:
```diff
@@ -0,0 +1,127 @@
+ */
+import { DEFAULT_CONTEXT } from './constants';
+
+/**
```
I really like how you document code. Just wanted to emphasize it as I look at this PR 💯 
How do you align @param definitions?

aduth replied:
>How do you align @param definitions?

I try to follow the guidelines in the [Documentation Standard](https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/) which in practice are to keep each fragment space-aligned (type, name, description) within each line-delimited grouping (param, return, etc).

gziolo replied:
I see, 80 characters per line is difficult to achieve but makes sense 👍

aduth replied:
I forgot the 80 character thing was part of the standard. It's self-imposed elsewhere. It becomes ugly and tempting to bypass in cases like this, but I think it's a motivator to try to avoid complexity (e.g. in the naming of the type).

gziolo commented on packages/notices/src/store/actions.js:
```diff
@@ -0,0 +1,127 @@
+ * @param {?Array<WPNoticeAction>} options.actions       User actions to be
+ *                                                       presented with notice.
+ */
+export function* createNotice( status = 'info', content, options = {} ) {
```
Do we have any docs about using generators with action creators? I would love to read it myself to better understand what is possible as of today. I feel like I'm missing some great capabilities by not knowing what options are provided out of the box.

aduth replied:
>Do we have any docs about using generators with action creators?

At the moment, I think they're best described by:

- https://github.com/WordPress/gutenberg/tree/master/packages/data#controls
- https://github.com/WordPress/gutenberg/tree/master/packages/redux-routine/

gziolo replied:
Thanks for sharing thos links, I will check them out on Monday.

gziolo commented on packages/editor/src/store/test/effects.js:
```diff
@@ -257,17 +268,7 @@ describe( 'effects', () => {
-				},
-				type: 'CREATE_NOTICE',
-			} ) );
+			expect( dataDispatch( 'core/notices' ).createSuccessNotice ).toHaveBeenCalled();
```
Is there any reason why we can assert using more detailed check similar to what was previousle and what I see later in the file, e.g.:
```js
expect( dataDispatch( 'core/notices' ).createErrorNotice ).toHaveBeenCalledWith( 'Publishing failed', { id: SAVE_POST_NOTICE_ID } );
```

aduth replied:
> Is there any reason why we can assert using more detailed check similar to what was previousle and what I see later in the file, e.g.:
> 
> ```js
> expect( dataDispatch( 'core/notices' ).createErrorNotice ).toHaveBeenCalledWith( 'Publishing failed', { id: SAVE_POST_NOTICE_ID } );
> ```

Updated in rebased c5985282b2a2b977fb10ddb37eb8a797fa13e566.

gziolo commented on packages/editor/src/store/effects/test/reusable-blocks.js:
```diff
@@ -37,6 +36,7 @@ import {
 	fetchReusableBlocks as fetchReusableBlocksAction,
 } from '../../actions';
 import reducer from '../../reducer';
+import '../../..';
```
Can we add a comment why we import this thing? It can be copy and paste from what we had before.

aduth replied:
> Can we add a comment why we import this thing? It can be copy and paste from what we had before.

Updated in rebased 009b8e916b3ae9fac32af65261713feab454097f.

gziolo commented on packages/editor/src/store/effects/reusable-blocks.js:
```diff
@@ -16,15 +16,17 @@ import {
 	cloneBlock,
 } from '@wordpress/blocks';
 import { __ } from '@wordpress/i18n';
+// TODO: Ideally this would be the only dispatch in scope. This requires either
```
I finally get the idea. It's not perfect but let's move on and refactor later.

@aduth @youknowriad how do you see the future or effects? Should we invest more time in improving them to make them capable of working with multiple stores?

gziolo replied:
I personally really like how speak is implemented. It’s very easy to follow when you omit all the fancy syntax that it requires 😅 I asked that because if we bet on controls then we don’t have to worry about using global dispatch hrre. It’s just optical change when we pass it as argument rather than store.

gziolo commented on packages/editor/src/store/effects/reusable-blocks.js:
```diff
@@ -131,13 +133,14 @@ export const saveReusableBlocks = async ( action, store ) => {
 		} );
 		const message = isTemporary ? __( 'Block created.' ) : __( 'Block updated.' );
-		dispatch( createSuccessNotice( message, { id: REUSABLE_BLOCK_NOTICE_ID } ) );
+		dataDispatch( 'core/notices' ).createSuccessNotice( message, {
```
It makes me wonder if we should refactor effects to take `data` as param rather than `store`. This would align closer to what we have in other parts of the applications, in particular, `withSelect` and `withDispatch`.

aduth replied:
> It makes me wonder if we should refactor effects to take `data` as param rather than `store`. 

By date, would you mean the current registry? I guess that would work as a solution for https://github.com/WordPress/gutenberg/pull/9617/files#r228623495, but as noted in that comment, we'll need a similar solution for selectors and controls as well.

gziolo replied:
Yes, this is what I thought.

gziolo commented on packages/editor/src/store/effects/posts.js:
```diff
@@ -120,8 +120,8 @@ export const requestPostUpdate = async ( action, store ) => {
-		dispatch( removeNotice( SAVE_POST_NOTICE_ID ) );
-		dispatch( removeNotice( AUTOSAVE_POST_NOTICE_ID ) );
+		dataDispatch( 'core/notices' ).removeNotice( SAVE_POST_NOTICE_ID );
+		dataDispatch( 'core/notices' ).removeNotice( 'autosave-exists' );
```
Why do we stop using this `AUTOSAVE_POST_NOTICE_ID` constant? Are there any blockers? Can't we use the trick with `SAVE_POST_NOTICE_ID`?

aduth replied:
I kinda feel these constants are a waste of our resources, in the same way we don't bother with action type constants. I'll see about at least making things consistent, even if that means reintroducing the constant.

aduth replied:
> Why do we stop using this `AUTOSAVE_POST_NOTICE_ID` constant? Are there any blockers? Can't we use the trick with `SAVE_POST_NOTICE_ID`?

Looking at this more, I don't really like the idea of importing store constants into the `EditorProvider` component. And in general, I don't really like how we've implemented this notion of notices which should be dismissed on the next save. Rather than hard-coding their removal, it seems like it might be better to have some separate state tracking IDs of notices which should be removed on the next save. This way the `EditorProvider` could add to the state at the same time it's creating its notice, keeping the ID colocated.

Thoughts? Maybe better to defer to a follow-up pull request, keeping this one as-is?

gziolo replied:
I’m on mobile so I can’t check the last part of your comment. It sounds like a good idea. We can discuss in the future. Let’s get this in first 😃

gziolo commented:
> There are local uncommitted changes after one or both of 'npm install' or 'npm run docs:build'!

Your own Travis check bites :)

gziolo commented:
To answer more of your questions:

> With introduction of "context", should we consider most of the notices as we have today as part of an "editor" context, or does the idea of a global context represent what we're showing today in the editor below the heading? It'd be my idea that what we have today serves the "global" usage, but it is screen-specific, and we might want to consider "properly global" notices (things like notifications center), where our notices aren't quite as global.

That's an interesting question. I would keep global for some time and see how it evolves. As far as I understand the place we render them is going to contain notices that as of today are considered global.

gziolo commented:
It works as expected with all the changes introduced. I can see all notices that were rendered before and I can use all links. I will give it 👍 as soon as Travis is happy and my questions are answered.

gziolo commented on lib/client-assets.php:
```diff
@@ -442,8 +442,8 @@ function gutenberg_register_scripts_and_styles() {
 			'wp-polyfill',
-			'wp-url',
 			'wp-rich-text',
+			'wp-url',
```
@atimmer @omarreiss @pento  - pinging you, so you aware that this is coming up soon and we will have to update Core, too.

gziolo replied:
> Just a quick note that we should notify the people working on Core integration because this impacts the scripts registration...

@youknowriad good call, done 😄

youknowriad replied:
There's also a new script to register...

atimmer replied:
Thanks for the notification, putting it on my list for when I update the packages based on Gutenberg 4.2.

aduth commented on packages/editor/src/store/actions.js:
```diff
@@ -788,3 +741,36 @@ export function disablePublishSidebar() {
+
+	dispatch( 'core/notices' ).createNotice( status, content, options );
+
+	return { type: '__INERT__' };
```
Realistically, it doesn't matter much, particularly for deprecations slated for removal, but I can plan to align for consistency's sake.

aduth replied:
I don't see an instance of `DO_NOTHING` from his pull requests, so I'm going to assume it's either not been merged or no longer relevant. If it's not been merged, then... I win if I get mine in first 😛

aduth committed:
Editor: Port deprecated notices action to module usage

aduth commented on packages/editor/src/store/test/effects.js:
```diff
@@ -257,17 +268,7 @@ describe( 'effects', () => {
-			expect( dispatch ).toHaveBeenCalledWith( expect.objectContaining( {
-				notice: {
-					content: <p>Post published.{ ' ' }<a>View post</a></p>, // eslint-disable-line jsx-a11y/anchor-is-valid
-					id: 'SAVE_POST_NOTICE_ID',
```
To my point about value of constants, we don't even use them consistently as constants.

gziolo replied:
Yes, let’s not use them at all then 😃

aduth commented:
Thanks so much for the detailed review of this behemoth @gziolo .

I've pushed a rebased branch with conflicts resolved, feedback addressed, and commits squashed (with attribution preserved).

gziolo reviewed (approved):
Thanks for addressing my feedback and your in-depth answers. Let’s 🚢 it as soon as Travis is happy about the lock file.

aduth commented:
Created two issues to track follow-up work:

- #11201 Data: Consolidate onSubKey (and other helpers) to data module
- #11202 Editor: Eliminate or refactor store notices constants
